win32ss/user/ntuser/message.c

```diff
@@ -459,8 +483,30 @@ CopyMsgToKernelMem(MSG *KernelModeMsg, MSG *UserModeMsg, PMSGMEMORY MsgMemoryEnt
         /* Copy data if required */
         if (0 != (MsgMemoryEntry->Flags & MMS_FLAG_READ))
         {
+            WCHAR lParamMsg[sizeof(StrUserKernel[0])/2 + 1] = { 0 };
+
             TRACE("Copy Message %u from usermode buffer\n", KernelModeMsg->message);
-            Status = MmCopyFromCaller(KernelMem, (PVOID) UserModeMsg->lParam, Size);
+
+            _SEH2_TRY
+            {
+                if (UserModeMsg->lParam)
+                    RtlCopyMemory( lParamMsg, (WCHAR*)UserModeMsg->lParam, sizeof(lParamMsg));
```

Comment:
```suggestion
                    RtlCopyMemory(lParamMsg, (WCHAR*)UserModeMsg->lParam, sizeof(lParamMsg));
```

Reply:
I'm not fully sure this can work reliably.
If someone passes an arbitrary pointer to a _**non-null-terminated string**_ via the lParam with a size completely unrelated to the "hardcoded" sizeof(...) .

Later down you call `PosInArray(lParamMsg)` and wcscpy outside a SEH2 block, assuming the copied string is properly NUL-terminated, but you haven't verified that this assumption is correct.

Reply:
If someone passes a bad value string as you suggest, when this goes to ```PosInArray``` then it will not match any of the 3 strings there are will return a -1 which will disallow it getting to ```wcscpy``` function. Maybe it can be extended later for more general strings. Thanks.

Reply:
Extra space removed. I still do not understand your reasoning. If you have a proposed solution or suggestion, please provide it. Thanks.